File: Assets/src/model/enemies/Jackal.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using UnityEngine;
// run fast, fear other jackals nearby when they die
[System.Serializable]
[ObjectInfo(description: "Alternates moving 1 and 2 tiles.\nRuns away when another Jackal dies.\nChases you.")]
public class Jackal : AIActor {
  public static new ActionCosts StaticActionCosts = new ActionCosts(Actor.StaticActionCosts) {
    [ActionType.MOVE] = 0.67f,
  };

  protected override ActionCosts actionCosts => Jackal.StaticActionCosts;
  public Jackal(Vector2Int pos) : base(pos) {
    faction = Faction.Enemy;
    hp = baseMaxHp = 2;
  }

  protected override ActorTask GetNextTask() {
    var player = GameModel.main.player;
    if (CanTargetPlayer()) {
      if (IsNextTo(player)) {
        return new AttackTask(this, player);
      } else {
        return new ChaseTargetTask(this, player);
      }
    } else {
      return new MoveRandomlyTask(this);
    }
  }

  public override void HandleDeath(Entity source) {
    base.HandleDeath(source);
    var jackalsToAlert = floor.ActorsInCircle(pos, 7).Where((actor) => (actor is Jackal || actor is JackalBoss) && floor.TestVisibility(pos, actor.pos)).ToList();
    GameModel.main.EnqueueEvent(() => {
      foreach (var jackal in jackalsToAlert) {
        jackal.SetTasks(new RunAwayTask(jackal, pos, 6));
      }
    });
  }

  internal override (int, int) BaseAttackDamage() => (1, 2);
}

[Serializable]
[ObjectInfo(description: "Summons jackals if there are none on the map.")]
public class JackalBoss : Boss {
  // moves faster than jackals
  public override float turnPriority => base.turnPriority - 1;

  public JackalBoss(Vector2Int pos) : base(pos) {
    hp = baseMaxHp = 16;
    faction = Faction.Enemy;
    ClearTasks();
  }

  public float cooldown = 0;
  public override float Step() {
    var dt = base.Step();
    if (cooldown > 0) {
      cooldown -= dt;
    }
    return dt;
  }

  public override void HandleDeath(Entity source) {
    base.HandleDeath(source);
    // kill all jackals on the map
    foreach (var j in AllJackals.ToArray()) {
      j.Kill(this);
    }
  }

  public IEnumerable<Jackal> AllJackals => floor.bodies.OfType<Jackal>();

  internal override (int, int) BaseAttackDamage() {
    return (2, 3);
  }

  protected override ActorTask GetNextTask() {
    var player = GameModel.main.player;
    bool shouldCast = cooldown <= 0 && AllJackals.Count() < 1;
    if (shouldCast) {
      return new TelegraphedTask(this, 1, new GenericBaseAction(this, SummonJackals));
    }
    if (CanTargetPlayer()) {
      if (IsNextTo(player)) {
        return new AttackTask(this, player);
      } else {
        var task = new ChaseTargetTask(this, player);
        // recompute GetNextTask for cooldown
        task.maxMoves = 1;
        return task;
      }
    } else {
      return new MoveRandomlyTask(this);
    }
  }

  int numToSummon = 2;

  public void SummonJackals() {
    cooldown = 9;
    var perimeter = floor.EnumeratePerimeter(1).ToList();
    perimeter.Shuffle();
    for (int i = 0; i < numToSummon; i++) {
      floor.Put(new Jackal(perimeter[i]));
    }
    numToSummon++;
  }
}

// [Serializable]
// [ObjectInfo("slime", description: "Deals 1 damage to any non-Blob that walks into it.\nRemoved when you walk into it, or the Blobmother dies.")]
// public class BlobSlime : Grass, IActorEnterHandler {
//   public BlobSlime(Vector2Int pos) : base(pos) {}

//   public void HandleActorEnter(Actor who) {
//     if (!(who is Blob || who is Blobmother || who is MiniBlob)) {
//       who.TakeDamage(1, this);
//       Kill(who);
//     }
//   }
// }

File: web/src/model/enemies/Jackal.ts
import { AIActor } from './AIActor';
import { ActorTask } from '../ActorTask';
import { AttackTask } from '../tasks/AttackTask';
import { ChaseTargetTask } from '../tasks/ChaseTargetTask';
import { MoveRandomlyTask } from '../tasks/MoveRandomlyTask';
import { RunAwayTask } from '../tasks/RunAwayTask';
import { WaitTask } from '../tasks/WaitTask';
import { TelegraphedTask } from '../tasks/TelegraphedTask';
import { ActionCosts, GenericBaseAction } from '../BaseAction';
import { ActionType, Faction } from '../../core/types';
import type { IDeathHandler } from '../../core/types';
import { Vector2Int } from '../../core/Vector2Int';
import { GameModelRef } from '../GameModelRef';
import { MyRandom } from '../../core/MyRandom';
import { entityRegistry } from '../../generator/entityRegistry';

const DEATH_HANDLER = Symbol.for('IDeathHandler');

/**
 * Fast enemy that flees when another Jackal dies nearby.
 * Port of C# Jackal.cs.
 */
export class Jackal extends AIActor implements IDeathHandler {
  readonly [DEATH_HANDLER] = true as const;

  private fastStepsRemaining = 2;
  private wasAdjacentAtTurnStart = false;

  get sprintReady(): boolean {
    return this.fastStepsRemaining > 0;
  }

  protected get actionCosts(): ActionCosts {
    return new ActionCosts([
      [ActionType.MOVE, this.fastStepsRemaining > 0 ? 0.5 : 1],
      [ActionType.ATTACK, 1],
      [ActionType.WAIT, 1],
      [ActionType.GENERIC, 1],
    ]);
  }

  constructor(pos: Vector2Int) {
    super(pos);
    this.faction = Faction.Enemy;
    this.hp = this._baseMaxHp = 1;
  }

  baseAttackDamage(): [number, number] {
    return [1, 1];
  }

  step(): number {
    if (this.fastStepsRemaining === 2) {
      this.wasAdjacentAtTurnStart = this.isNextTo(GameModelRef.main.player);
    }
    const cost = super.step();
    if (cost <= 0.5) {
      this.fastStepsRemaining--;
    } else {
      this.fastStepsRemaining = 2;
    }
    return cost;
  }

  protected getNextTask(): ActorTask {
    const player = GameModelRef.main.player;
    if (this.canTargetPlayer()) {
      if (this.isNextTo(player)) {
        if (this.sprintReady && !this.wasAdjacentAtTurnStart) {
          return new WaitTask(this, 1);
        }
        return new AttackTask(this, player);
      }
      return new ChaseTargetTask(this, player);
    }
    return new MoveRandomlyTask(this);
  }

  handleDeath(_source: any): void {
    const floor = this.floor;
    if (!floor) return;
    const myPos = this.pos;

    // Alert nearby jackals/jackal bosses within radius 7 that can see this jackal
    const toAlert: AIActor[] = [];
    for (const pos of floor.enumerateCircle(myPos, 7)) {
      const body = floor.bodies.get(pos);
      if (body && (body instanceof Jackal || body instanceof JackalBoss)) {
        if (body !== this && floor.testVisibility(myPos, body.pos)) {
          toAlert.push(body as AIActor);
        }
      }
    }

    GameModelRef.main.enqueuEvent(() => {
      for (const jackal of toAlert) {
        (jackal as any).setTasks(new RunAwayTask(jackal, myPos, 6));
      }
    });
  }
}

/**
 * Boss that summons jackals when none remain.
 * Port of C# JackalBoss (extends Boss → AIActor).
 */
export class JackalBoss extends AIActor implements IDeathHandler {
  readonly [DEATH_HANDLER] = true as const;

  get turnPriority(): number {
    return super.turnPriority - 1;
  }

  private cooldown = 0;
  private numToSummon = 2;

  constructor(pos: Vector2Int) {
    super(pos);
    this.faction = Faction.Enemy;
    this.hp = this._baseMaxHp = 16;
    this.clearTasks();
  }

  baseAttackDamage(): [number, number] {
    return [2, 3];
  }

  step(): number {
    const dt = super.step();
    if (this.cooldown > 0) {
      this.cooldown -= dt;
    }
    return dt;
  }

  handleDeath(_source: any): void {
    // Kill all jackals on the map when the boss dies
    const jackals = this.allJackals();
    for (const j of jackals) {
      j.kill(this);
    }
  }

  private allJackals(): Jackal[] {
    if (!this.floor) return [];
    return this.floor.bodies.where(b => b instanceof Jackal) as Jackal[];
  }

  protected getNextTask(): ActorTask {
    const player = GameModelRef.main.player;
    const shouldCast = this.cooldown <= 0 && this.allJackals().length < 1;

    if (shouldCast) {
      return new TelegraphedTask(
        this,
        1,
        new GenericBaseAction(this, () => this.summonJackals()),
      );
    }

    if (this.canTargetPlayer()) {
      if (this.isNextTo(player)) {
        return new AttackTask(this, player);
      }
      const chase = new ChaseTargetTask(this, player);
      chase.maxMoves = 1;
      return chase;
    }
    return new MoveRandomlyTask(this);
  }

  private summonJackals(): void {
    this.cooldown = 9;
    const perimeter = [...this.floor!.enumeratePerimeter(1)];
    MyRandom.Shuffle(perimeter);
    for (let i = 0; i < this.numToSummon && i < perimeter.length; i++) {
      this.floor!.put(new Jackal(perimeter[i]));
    }
    this.numToSummon++;
  }
}

entityRegistry.register('Jackal', Jackal);
entityRegistry.register('JackalBoss', JackalBoss);
